### PR TITLE
snapshots: Allow disk-only snapshots of running machines

### PR DESCRIPTION
**Pull request by mvollmer**
- [x] Use "Full system" instead of "Complete". That's the libvirt terminology.
- [x] Detect guest agent and add "quiesce" flag is found.
- [x] Test quiesce functionality.... Maybe switch from cirros to alpine, which has qemu-guest-agent

image

image

image

### CODE REVIEW

**Comment by mvollmer** [comment]
I wasn't able to revert to a disk-only snapshot:

image

Probably tracked by https://issues.redhat.com/browse/RHEL-21549 and fixed by https://www.mail-archive.com/devel@lists.libvirt.org/msg01703.html, but not yet in rhel-9-4.

Should this feature here go to rhel-9-4 or be put on hold until rhel-9-5?

**Comment by mvollmer** [comment]
> Should this feature here go to rhel-9-4 or be put on hold until rhel-9-5?

Alright, current "main" goes to rhel-9-5 already, this is good to go.

**Comment by mvollmer** [comment]
I should've used the term "disk-only" instead of "partial" in the code, but I figured out too late that that's what libvirt calls them.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 1]
```diff
@@ -51,6 +56,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
+                <Radio isChecked={!isPartial}
+                       id="snapshot-create-dialog-type-complete"
+                       onChange={() => onValueChanged("isPartial", false)}
+                       label={_("Complete")} />
+                <Radio isChecked={isPartial}
+                       id="snapshot-create-dialog-type-partial"
+                       onChange={() => onValueChanged("isPartial", true)}
+                       label={_("Disk only")} />
+            </Split>
+        </FormGroup>
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:90]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 2]
```diff
@@ -138,7 +179,7 @@ export class CreateSnapshotModal extends React.Component {
         else if (!name)
             validationError.name = _("Name can not be empty");
 
-        if (isExternal && vm.state === "running" && !memoryPath)
+        if (isExternal && vm.state === "running" && !isPartial && !memoryPath)
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:182]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 3]
```diff
@@ -178,16 +219,18 @@ export class CreateSnapshotModal extends React.Component {
...
+        const { name, description, isPartial, memoryPath } = this.state;
         const validationError = this.onValidate();
 
         const body = (
             <Form onSubmit={e => e.preventDefault()} isHorizontal>
                 <NameRow name={name} validationError={validationError.name} onValueChanged={this.onValueChanged} />
-                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
                 {isExternal && vm.state === 'running' &&
+                    <TypeRow isPartial={isPartial} onValueChanged={this.onValueChanged} />}
+                {isExternal && vm.state === 'running' && !isPartial &&
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:230]: These 2 added lines are not executed by any test.

**Comment by mvollmer** [comment]
> I should've used the term "disk-only" instead of "partial" in the code, but I figured out too late that that's what libvirt calls them.

Fixed.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 4]
```diff
@@ -51,6 +56,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
+                <Radio isChecked={!isDiskOnly}
+                       id="snapshot-create-dialog-type-complete"
+                       onChange={() => onValueChanged("isDiskOnly", false)}
+                       label={_("Complete")} />
+                <Radio isChecked={isDiskOnly}
+                       id="snapshot-create-dialog-type-disk-only"
+                       onChange={() => onValueChanged("isDiskOnly", true)}
+                       label={_("Disk only")} />
+            </Split>
+        </FormGroup>
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:90]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 5]
```diff
@@ -138,7 +179,7 @@ export class CreateSnapshotModal extends React.Component {
         else if (!name)
             validationError.name = _("Name can not be empty");
 
-        if (isExternal && vm.state === "running" && !memoryPath)
+        if (isExternal && vm.state === "running" && !isDiskOnly && !memoryPath)
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:182]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 6]
```diff
@@ -178,16 +219,18 @@ export class CreateSnapshotModal extends React.Component {
...
+        const { name, description, isDiskOnly, memoryPath } = this.state;
         const validationError = this.onValidate();
 
         const body = (
             <Form onSubmit={e => e.preventDefault()} isHorizontal>
                 <NameRow name={name} validationError={validationError.name} onValueChanged={this.onValueChanged} />
-                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
                 {isExternal && vm.state === 'running' &&
+                    <TypeRow isDiskOnly={isDiskOnly} onValueChanged={this.onValueChanged} />}
+                {isExternal && vm.state === 'running' && !isDiskOnly &&
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:230]: These 2 added lines are not executed by any test.

**Comment by mvollmer** [comment]
This is blocked on https://issues.redhat.com/browse/RHEL-21549

Cockpit should only offer the creation of disk-only external snapshots when they can also be restored. We might have to detect that at run-time, maybe by looking at the libvirt version.

**Comment by martinpitt** [comment]
For the record: I will never approve this, this is a dangerous and crazy practice to offer and train admins to do. I won't make a fuss if someone else approves, but I refuse any responsibility for this, sorry.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 7]
```diff
@@ -51,6 +56,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
+                       id="snapshot-create-dialog-type-full"
+                       onChange={() => onValueChanged("isDiskOnly", false)}
+                       label={_("Full system")} />
+                <Radio isChecked={isDiskOnly}
+                       id="snapshot-create-dialog-type-disk-only"
+                       onChange={() => onValueChanged("isDiskOnly", true)}
+                       label={_("Disk only")} />
+            </Split>
+        </FormGroup>
+    );
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:91]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 8]
```diff
@@ -51,6 +56,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:91]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 9]
```diff
@@ -138,7 +179,7 @@ export class CreateSnapshotModal extends React.Component {
         else if (!name)
             validationError.name = _("Name can not be empty");
 
-        if (isExternal && vm.state === "running" && !memoryPath)
+        if (isExternal && vm.state === "running" && !isDiskOnly && !memoryPath)
             validationError.memory = _("Memory file can not be empty");
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:183]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 10]
```diff
@@ -178,16 +219,18 @@ export class CreateSnapshotModal extends React.Component {
...
             <Form onSubmit={e => e.preventDefault()} isHorizontal>
                 <NameRow name={name} validationError={validationError.name} onValueChanged={this.onValueChanged} />
-                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
                 {isExternal && vm.state === 'running' &&
+                    <TypeRow isDiskOnly={isDiskOnly} onValueChanged={this.onValueChanged} />}
+                {isExternal && vm.state === 'running' && !isDiskOnly &&
                     <MemoryPathRow memoryPath={memoryPath} onValueChanged={this.onValueChanged}
                                    validationError={validationError.memory} />}
+                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
             </Form>
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:234]: These 2 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 11]
```diff
@@ -52,6 +57,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
+                <Radio isChecked={!isDiskOnly}
+                       id="snapshot-create-dialog-type-full"
+                       onChange={() => onValueChanged("isDiskOnly", false)}
+                       label={_("Full system")} />
+                <Radio isChecked={isDiskOnly}
+                       id="snapshot-create-dialog-type-disk-only"
+                       onChange={() => onValueChanged("isDiskOnly", true)}
+                       label={_("Disk only")} />
+            </Split>
+        </FormGroup>
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:91]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 12]
```diff
@@ -139,30 +180,26 @@ export class CreateSnapshotModal extends React.Component {
         else if (!name)
             validationError.name = _("Name can not be empty");
 
-        if (isExternal && vm.state === "running" && !memoryPath)
+        if (isExternal && vm.state === "running" && !isDiskOnly && !memoryPath)
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:183]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 13]
```diff
@@ -139,30 +180,26 @@ export class CreateSnapshotModal extends React.Component {
...
-                connectionName: vm.connectionName,
-                vmId: vm.id,
+                vm,
                 name,
                 description,
-                memoryPath: vm.state === "running" && memoryPath,
-                disks,
                 isExternal,
-                storagePools
+                memoryPath: isExternal && vm.state === "running" && !isDiskOnly && memoryPath,
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:202]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 14]
```diff
@@ -182,16 +219,18 @@ export class CreateSnapshotModal extends React.Component {
...
+        const { name, description, isDiskOnly, memoryPath } = this.state;
         const validationError = this.onValidate();
 
         const body = (
             <Form onSubmit={e => e.preventDefault()} isHorizontal>
                 <NameRow name={name} validationError={validationError.name} onValueChanged={this.onValueChanged} />
-                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
                 {isExternal && vm.state === 'running' &&
+                    <TypeRow isDiskOnly={isDiskOnly} onValueChanged={this.onValueChanged} />}
+                {isExternal && vm.state === 'running' && !isDiskOnly &&
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:230]: These 2 added lines are not executed by any test.

File: src/libvirt-xml-create.js [hunk 15]
```diff
@@ -253,36 +253,11 @@ export function getSnapshotXML(name, description, disks, memoryPath, isExternal,
...
-            diskElem.setAttribute('name', disk.target);
-            diskElem.setAttribute('snapshot', 'external');
-            disksElem.appendChild(diskElem);
-        });
-        snapElem.appendChild(disksElem);
+    if (memoryPath) {
+        const memoryElem = doc.createElement('memory');
+        memoryElem.setAttribute('snapshot', 'external');
+        memoryElem.setAttribute('file', memoryPath);
+        snapElem.appendChild(memoryElem);
```
Comment (cockpituous) [inline src/libvirt-xml-create.js:260]: These 5 added lines are not executed by any test.

File: src/libvirtApi/snapshot.js [hunk 16]
```diff
@@ -29,11 +29,37 @@ import { parseDomainSnapshotDumpxml } from '../libvirt-xml-parse.js';
...
     const flags = (!isExternal || memoryPath) ? 0 : Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY;
-    const xmlDesc = getSnapshotXML(name, description, disks, memoryPath, isExternal, storagePools, connectionName);
-    return call(connectionName, vmId, 'org.libvirt.Domain', 'SnapshotCreateXML', [xmlDesc, flags], { timeout, type: 'su' });
+    const xmlDesc = getSnapshotXML(name, description, memoryPath);
+
+    // We really don't want to make disk only snapshots of running
+    // machines by accident.
+
+    if (vm.state === "running" && (flags & Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY))
+        throw new Error("Cowardly refusing to make a disk-only snapshot of a running machine");
```
Comment (cockpituous) [inline src/libvirtApi/snapshot.js:59]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx (hunk 11 above)
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:91]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 17]
```diff
@@ -139,30 +180,27 @@ export class CreateSnapshotModal extends React.Component {
         else if (!name)
             validationError.name = _("Name can not be empty");
 
-        if (isExternal && vm.state === "running" && !memoryPath)
+        if (isExternal && vm.state === "running" && !isDiskOnly && !memoryPath)
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:183]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 18]
```diff
@@ -139,30 +180,27 @@ export class CreateSnapshotModal extends React.Component {
...
-                vmId: vm.id,
+                vm,
                 name,
                 description,
-                memoryPath: vm.state === "running" && memoryPath,
-                disks,
                 isExternal,
-                storagePools
+                doQuiesce: vm.hasAgent,
+                memoryPath: isExternal && vm.state === "running" && !isDiskOnly && memoryPath,
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:203]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 19]
```diff
@@ -182,16 +220,18 @@ export class CreateSnapshotModal extends React.Component {
...
+        const { name, description, isDiskOnly, memoryPath } = this.state;
         const validationError = this.onValidate();
 
         const body = (
             <Form onSubmit={e => e.preventDefault()} isHorizontal>
                 <NameRow name={name} validationError={validationError.name} onValueChanged={this.onValueChanged} />
-                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
                 {isExternal && vm.state === 'running' &&
+                    <TypeRow isDiskOnly={isDiskOnly} onValueChanged={this.onValueChanged} />}
+                {isExternal && vm.state === 'running' && !isDiskOnly &&
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:231]: These 2 added lines are not executed by any test.

File: src/libvirt-xml-parse.js [hunk 20]
```diff
@@ -563,6 +565,22 @@ function parseDumpxmlForTPM(devicesElem) {
...
+    const channelElems = devicesElem.getElementsByTagName('channel');
+    for (let i = 0; i < channelElems.length; ++i) {
+        const channel = channelElems[i];
+        const target = getSingleOptionalElem(channel, "target");
+
+        if (target &&
+            target.getAttribute("type") == "virtio" &&
+            target.getAttribute("name") == "org.qemu.guest_agent.0" &&
+            target.getAttribute("state") == "connected")
+            return true;
```
Comment (cockpituous) [inline src/libvirt-xml-parse.js:578]: This added line is not executed by any test.

File: src/libvirtApi/snapshot.js [hunk 21]
```diff
@@ -29,11 +29,34 @@ import { parseDomainSnapshotDumpxml } from '../libvirt-xml-parse.js';
...
+    // --------------------------------------------------------------------
+    // shutoff   |  false      |  null          =>  internal
+    // shutoff   |  true       |  null          =>  external
+    // running   |  false      |  null          =>  internal full system
+    // running   |  false      |  non-null      =>  external full system
+    // running   |  true       |  null          =>  external disk only
+    //
+    // Other cases are errors.
+
+    let flags = (!isExternal || memoryPath) ? 0 : Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY;
```
Comment (cockpituous) [inline src/libvirtApi/snapshot.js:52]: This added line is not executed by any test.

File: src/libvirtApi/snapshot.js [hunk 22]
```diff
@@ -29,11 +29,34 @@ import { parseDomainSnapshotDumpxml } from '../libvirt-xml-parse.js';
...
+    const xmlDesc = getSnapshotXML(name, description, memoryPath);
+
+    if (vm.state === "running" && (flags & Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY) && doQuiesce)
+        flags |= Enum.VIR_DOMAIN_SNAPSHOT_CREATE_QUIESCE;
```
Comment (cockpituous) [inline src/libvirtApi/snapshot.js:56]: These 2 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx (hunk 11 above)
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:91]: These 32 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 23]
```diff
@@ -139,7 +180,7 @@ export class CreateSnapshotModal extends React.Component {
         else if (!name)
             validationError.name = _("Name can not be empty");
 
-        if (isExternal && vm.state === "running" && !memoryPath)
+        if (isExternal && vm.state === "running" && !isDiskOnly && !memoryPath)
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:183]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 24]
```diff
@@ -158,7 +199,8 @@ export class CreateSnapshotModal extends React.Component {
                 name,
                 description,
                 isExternal,
-                memoryPath: isExternal && vm.state === "running" && memoryPath,
+                doQuiesce: vm.hasAgent,
+                memoryPath: isExternal && vm.state === "running" && !isDiskOnly && memoryPath,
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:203]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 25]
```diff
@@ -178,16 +220,18 @@ export class CreateSnapshotModal extends React.Component {
...
+        const { name, description, isDiskOnly, memoryPath } = this.state;
         const validationError = this.onValidate();
 
         const body = (
             <Form onSubmit={e => e.preventDefault()} isHorizontal>
                 <NameRow name={name} validationError={validationError.name} onValueChanged={this.onValueChanged} />
-                <DescriptionRow description={description} onValueChanged={this.onValueChanged} />
                 {isExternal && vm.state === 'running' &&
+                    <TypeRow isDiskOnly={isDiskOnly} onValueChanged={this.onValueChanged} />}
+                {isExternal && vm.state === 'running' && !isDiskOnly &&
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:231]: These 2 added lines are not executed by any test.

File: src/libvirt-xml-parse.js (hunk 20 above)
Comment (cockpituous) [inline src/libvirt-xml-parse.js:578]: This added line is not executed by any test.

File: src/libvirtApi/snapshot.js [hunk 26]
```diff
@@ -49,14 +49,11 @@ export async function snapshotCreate({ vm, name, description, isExternal, memory
     //
     // Other cases are errors.
 
-    const flags = (!isExternal || memoryPath) ? 0 : Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY;
+    let flags = (!isExternal || memoryPath) ? 0 : Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY;
```
Comment (cockpituous) [inline src/libvirtApi/snapshot.js:52]: This added line is not executed by any test.

File: src/libvirtApi/snapshot.js [hunk 27]
```diff
@@ -49,14 +49,11 @@ export async function snapshotCreate({ vm, name, description, isExternal, memory
...
     const xmlDesc = getSnapshotXML(name, description, memoryPath);
 
-    // We really don't want to make disk only snapshots of running
-    // machines by accident.
-
-    if (vm.state === "running" && (flags & Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY))
-        throw new Error("Cowardly refusing to make a disk-only snapshot of a running machine");
+    if (vm.state === "running" && (flags & Enum.VIR_DOMAIN_SNAPSHOT_CREATE_DISK_ONLY) && doQuiesce)
+        flags |= Enum.VIR_DOMAIN_SNAPSHOT_CREATE_QUIESCE;
```
Comment (cockpituous) [inline src/libvirtApi/snapshot.js:56]: These 2 added lines are not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 28]
```diff
@@ -52,6 +57,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
+                                   <FlexItem>
+                                       <h4 className="popover-headline">{_("Full system")}</h4>
+                                       <p>{_("A full system snapshot can be used to revert the virtual machine to its exact current state where all its processes have the same run-time state as now. It might require significant extra storage to save the memory state.")}</p>
+                                   </FlexItem>
+                                   <FlexItem>
+                                       <h4 className="popover-headline">{_("Disk only")}</h4>
+                                       <p>{_("This type of snapshot only saves the storage devices of the virtual machine, and not the state of running processes.")}</p>
+                                   </FlexItem>
+                               </Flex>}>
+                           <button onClick={e => e.preventDefault()} className="pf-v5-c-form__group-label-help">
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:77]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx [hunk 29]
```diff
@@ -52,6 +57,41 @@ const NameRow = ({ onValueChanged, name, validationError }) => {
...
+                               <InfoAltIcon />
+                           </button>
+                       </Popover>}>
+            <Split hasGutter>
+                <Radio isChecked={!isDiskOnly}
+                       id="snapshot-create-dialog-type-full"
+                       onChange={() => onValueChanged("isDiskOnly", false)}
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:84]: This added line is not executed by any test.

File: src/libvirt-xml-parse.js (hunk 20 above)
Comment (cockpituous) [inline src/libvirt-xml-parse.js:578]: This added line is not executed by any test.

File: src/libvirtApi/snapshot.js (hunk 27 above)
Comment (cockpituous) [inline src/libvirtApi/snapshot.js:56]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx (hunk 28 above)
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:77]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx (hunk 29 above)
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:84]: This added line is not executed by any test.

File: src/libvirtApi/domain.js [hunk 30]
```diff
@@ -689,6 +689,16 @@ export async function domainGet({
...
 }
 
+export async function domainGetNow(vm, refresh) {
+    if (refresh)
+        await domainGet(vm);
+    for (const new_vm of store.getState().vms) {
+        if (new_vm.connectionName == vm.connectionName && new_vm.id == vm.id)
+            return new_vm;
+    }
+    return null;
```
Comment (cockpituous) [inline src/libvirtApi/domain.js:699]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
> I wasn't able to revert to a disk-only snapshot:

Hmm, reverting disk-only snapshots might not be that important. Maybe they are only used as a mechanism to get a stable view of disks for the purposes of backup. If that is the case, I would argue that Cockpit doesn't need to offer disk-only snapshots. Completing the backup requires a deep understanding of how external snapshots are implemented and this is best done on the command line or from a program.

**Comment by mvollmer** [comment]
> Completing the backup requires a deep understanding of how external snapshots are implemented and this is best done on the command line or from a program.

What Cockpit could do is to expose the snapshotted disks as NBD devices so that people can then copy them to their backups.

**Review by garrett — CHANGES_REQUESTED** [review]
I think this is very dangerous, both for stateful live snapshots (with state and memory) and disk-only, as it breaks expectations, which means it'll break someone's backups, which means someone using Cockpit will eventually lose their data.

We need to talk about snapshots holistically (the complete picture, including internal, external disk-only, and external live snapshots and the supportability and purpose of each) and work on designs that might help to mitigate the very real and likely risk for catastrophic data loss due to external snapshots.

**Comment by mac2net** [comment]
I mentioned on another thread here that I in fact did break a VM fiddling with the snapshots. I received additional instructions and will do another test shortly.
I understand that documentation has been added to the Redhat 9.4 release but I made a suggestion that documentation could be added via the GUI with a popup link that could at a minimum provide a link to documentation on the web,
image

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx (hunk 28 above)
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:77]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCreateModal.jsx (hunk 29 above)
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCreateModal.jsx:84]: This added line is not executed by any test.

File: src/libvirtApi/domain.js (hunk 30 above)
Comment (cockpituous) [inline src/libvirtApi/domain.js:699]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
This is dead for now. No disk-only snapshots of running VMs in Cockpit, see https://issues.redhat.com/browse/COCKPIT-1112

**Pull request closed without merging**